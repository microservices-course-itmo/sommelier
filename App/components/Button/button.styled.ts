import styled from 'styled-components/native'
import theme from '../../theme'

type ButtonProps = {
  width?: number
  borderRadius?: number
  height?: number
}

type UnderscoreButtonProps = {
  width?: number
  height?: number
  borderRadius?: number
  shadowWidth?: number
  shadowHeight?: number
  shadowOpacity?: number
  top?: number
  shadowColor?: keyof typeof theme.colors
}

type IconProps = {
  width?: number
  height?: number
  margin?: number
}

export const ButtonContainer = styled.TouchableOpacity<ButtonProps>`
  position: relative;
  background-color: ${({ theme }) => theme.colors.primaryRed};
  border-radius: ${({ borderRadius }) => borderRadius || 0};
  width: ${({ width }) => width || 0};
  height: ${({ height }) => height || 0};
  z-index: 2;
  margin: auto;
`

export const UnderscoreButtonContainer = styled.View<UnderscoreButtonProps>`
  position: absolute;
  margin-left: 30px;
  top: ${({ top }) => top || 0};
  background-color: ${({ theme }) => theme.colors.primaryDarkRed};
  border-radius: ${({ borderRadius }) => borderRadius || 0};
  width: ${({ width }) => width || 0};
  height: ${({ height }) => height || 0};
  shadow-offset: {
    width: ${({ shadowWidth }) => shadowWidth || 0};
    height: ${({ shadowHeight }) => shadowHeight || 0};
  }
  shadow-opacity: ${({ shadowOpacity }) => shadowOpacity || 0};
  shadow-color: ${({ theme }) => theme.colors.shadowButton};
  z-index: 1;
`

export const ButtonTextContainer = styled.View`
  flex: 1;
  flex-direction: row;
  justify-content: center;
  align-items: center;
`

export const Wrapper = styled.View`
  position: relative;
`

export const IconContainer = styled.View<IconProps>`
  width: ${({ width }) => width || 0};
  height: ${({ height }) => height || 0};

  margin-right: 9px;
`
