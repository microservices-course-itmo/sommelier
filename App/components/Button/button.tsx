import React, { ReactNode } from 'react'
import * as Styled from './button.styled'

type Props = {
  onPress: () => void
  children: ReactNode
}

const Button = ({ onPress = () => {}, children = '' }: Props) => (
  <Styled.ButtonContainer onPress={onPress} testID='sommelier-button'>
    <Styled.ButtonText>{children}</Styled.ButtonText>
  </Styled.ButtonContainer>
)

export default Button
