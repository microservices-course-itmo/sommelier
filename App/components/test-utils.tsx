import React from 'react'
import { render } from '@testing-library/react-native'
import { ThemeProvider } from 'styled-components'
import theme from '../theme'

const themeRender = (component: any) => {
  return render(<ThemeProvider theme={theme}>{component}</ThemeProvider>)
}

export * from '@testing-library/react-native'

export { themeRender }
