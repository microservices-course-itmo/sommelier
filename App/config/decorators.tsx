import { ThemeProvider } from 'styled-components'
import React from 'react'
import styled from 'styled-components/native'
import { theme } from '../theme'

const StyledBufferView = styled.View`
  flex: 1;
  margin-vertical: 40;
  margin-horizontal: 20;
`

export const ThemeDecorator = (storyFn: any) => (
  <ThemeProvider theme={theme}>{storyFn()}</ThemeProvider>
)

export const BufferView = (storyFn: any) => (
  <StyledBufferView>{storyFn()}</StyledBufferView>
)
