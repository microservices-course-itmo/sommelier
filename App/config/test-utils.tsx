import React, { ReactNode } from 'react'
import { render } from '@testing-library/react-native'
import { ThemeProvider } from 'styled-components/native'
import { theme } from '../theme'

const themeRender = (component: ReactNode) => {
  return render(<ThemeProvider theme={theme}>{component}</ThemeProvider>)
}

export * from '@testing-library/react-native'

export { themeRender }
