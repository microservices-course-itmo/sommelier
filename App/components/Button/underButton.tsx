import React from 'react'
import * as Styled from './button.styled'
import { theme } from '../../theme'
import { COLOR_KEYS } from '../../theme/colors'

type Props = {
  width?: number
  height?: number
  borderRadius?: number
  shadowWidth?: number
  shadowHeight?: number
  shadowOpacity?: number
  shadowColor?: keyof typeof theme.colors
  top?: number
}

const UnderscoreButton = ({
  width,
  height,
  borderRadius,
  shadowWidth,
  shadowHeight,
  shadowOpacity,
  top,
  shadowColor = COLOR_KEYS.shadowButton,
}: Props) => (
  <Styled.UnderscoreButtonContainer
    width={width}
    height={height}
    borderRadius={borderRadius}
    shadowWidth={shadowWidth}
    shadowHeight={shadowHeight}
    shadowOpacity={shadowOpacity}
    top={top}
    shadowColor={shadowColor}
  />
)

export default UnderscoreButton
