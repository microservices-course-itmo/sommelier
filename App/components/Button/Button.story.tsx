import React from 'react'
import { storiesOf } from '@storybook/react-native'
import { action } from '@storybook/addon-actions'
import { number, text, select } from '@storybook/addon-knobs'
import * as Styled from './button.styled'
import { FONT_SIZE_KEYS } from '../../theme/fontSizes/index'
import { COLOR_KEYS } from '../../theme/colors/index'

import Button from './button'
import UnderscoreButton from './underButton'
import { BufferView, ThemeDecorator } from '../../config/decorators'
import ButtonIcon from '../../../assets/images/ButtonIcon.svg'

const defaultButton = () => (
  <Styled.Wrapper>
    <Button
      onPress={action('tapper-default')}
      height={number('height', 51)}
      width={number('width', 327)}
      borderRadius={number('border-radius', 4)}
      text={text('text', 'Применить')}
      fontSize={select('font-size', FONT_SIZE_KEYS, FONT_SIZE_KEYS.hg)}
      color={select('text color', COLOR_KEYS, COLOR_KEYS.white)}
      iconHeight={number('icon height', 16)}
      iconWidth={number('icon width', 16)}
      icon={<ButtonIcon />}
    />
    <UnderscoreButton
      height={number('height 2', 35)}
      width={number('width 2', 311)}
      borderRadius={number('border-radius 2', 10)}
      shadowOpacity={number('shadow-opacity', 4)}
      shadowHeight={number('shadow-height', 4)}
      shadowWidth={number('shadow-width', 0)}
      top={number('top', 21)}
      shadowColor={select('shadow-color', COLOR_KEYS, COLOR_KEYS.shadowButton)}
    />
  </Styled.Wrapper>
)

storiesOf('Button', module)
  .addDecorator(BufferView)
  .addDecorator(ThemeDecorator)
  .add('Default button', defaultButton)
