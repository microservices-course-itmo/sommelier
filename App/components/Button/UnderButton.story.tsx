import React from 'react'
import { storiesOf } from '@storybook/react-native'
import { number, select } from '@storybook/addon-knobs'
import { COLOR_KEYS } from '../../theme/colors/index'

import UnderButton from './underButton'
import { BufferView, ThemeDecorator } from '../../config/decorators'

const defaultUnderButton = () => (
  <UnderButton
    height={number('height', 35)}
    width={number('width', 311)}
    borderRadius={number('border-radius', 10)}
    shadowOpacity={number('shadow-opacity', 4)}
    shadowHeight={number('shadow-height', -4)}
    shadowWidth={number('shadow-width', 0)}
    top={number('top', 0)}
    shadowColor={select('shadow-color', COLOR_KEYS, COLOR_KEYS.shadowButton)}
  />
)

storiesOf('UnderButton', module)
  .addDecorator(BufferView)
  .addDecorator(ThemeDecorator)
  .add('Default button', defaultUnderButton)
