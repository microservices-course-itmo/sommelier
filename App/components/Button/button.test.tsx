import React from 'react'
import { themeRender, fireEvent } from '../../config/test-utils'
import Button from './button'

describe('Button component', () => {
  const onPressMock = jest.fn()
  it('Component with only necessary props matches snapshot', () => {
    const component = themeRender(<Button onPress={onPressMock}>abc</Button>)
    expect(component.toJSON()).toMatchSnapshot()
  })
  it('OnPress function should be called', () => {
    const component = themeRender(<Button onPress={onPressMock}>abc</Button>)
    const button = component.getByTestId('sommelier-button')
    fireEvent.press(button)
    expect(onPressMock).toHaveBeenCalled()
  })
})
