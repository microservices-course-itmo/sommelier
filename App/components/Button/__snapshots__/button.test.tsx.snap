// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component Component with only necessary props matches snapshot 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#0265d6",
      "borderColor": "#0265d6",
      "borderRadius": 6,
      "borderWidth": 1,
      "marginVertical": 7,
      "opacity": 1,
      "paddingVertical": 14,
    }
  }
  testID="sommelier-button"
>
  <Text
    style={
      Array [
        Object {
          "alignSelf": "center",
          "color": "#fff",
          "fontSize": 20,
          "fontWeight": "500",
        },
      ]
    }
  >
    abc
  </Text>
</View>
`;
